types/resolve/index.d.ts

```diff
@@ -113,32 +113,48 @@ declare namespace resolve {
     preserveSymlinks?: boolean;
   }
 
-  export interface AsyncOpts extends Opts {
-    /** how to read files asynchronously (defaults to fs.readFile) */
-    readFile?: (file: string, cb: readFileCallback) => void;
+  interface BaseAsyncOpts extends Opts {
     /** function to asynchronously test whether a file exists */
     isFile?: (file: string, cb: existsCallback) => void;
     /** function to asynchronously test whether a directory exists */
     isDirectory?: (directory: string, cb: existsCallback) => void;
     /** function to asynchronously resolve a potential symlink to its real path */
     realpath?: (file: string, cb: realpathCallback) => void;
+  }
+
+  interface AsyncOptsWithReadFile extends BaseAsyncOpts {
+    /** how to read files asynchronously (defaults to fs.readFile) */
+    readFile?: (file: string, cb: readFileCallback) => void;
+  }
+
+  interface AsyncOptsWithReadPackage extends BaseAsyncOpts {
     /** function to asynchronously read and parse a package.json file */
     readPackage?: (readFile: (file: string, cb: readFileCallback) => void, pkgfile: string, cb: readPackageCallback) => void;
   }
 
-  export interface SyncOpts extends Opts {
-    /** how to read files synchronously (defaults to fs.readFileSync) */
-    readFileSync?: (file: string) => string | Buffer;
+  export type AsyncOpts = AsyncOptsWithReadFile | AsyncOptsWithReadPackage;
```

Comment:
not sure why these are interfaces versus types like `alwaysOpts & (readFileOpt | readPackageOpt)` - is there an advantage to using interfaces?

Reply:
It'll be named properly in the IDE (I _think_). Your version is better than having to resort to `never` though, so I can try that

Reply:
Didn't work

```diff
diff --git c/types/resolve/index.d.ts w/types/resolve/index.d.ts
index 102160c388..342bbdc4bf 100644
--- c/types/resolve/index.d.ts
+++ w/types/resolve/index.d.ts
@@ -122,21 +122,13 @@ declare namespace resolve {
     realpath?: (file: string, cb: realpathCallback) => void;
   }
 
-  interface AsyncOptsWithReadFile extends BaseAsyncOpts {
+  export type AsyncOpts = BaseAsyncOpts & ({
     /** how to read files asynchronously (defaults to fs.readFile) */
     readFile?: (file: string, cb: readFileCallback) => void;
-    /** function to asynchronously read and parse a package.json file */
-    readPackage?: never;
-  }
-
-  interface AsyncOptsWithReadPackage extends BaseAsyncOpts {
-    /** how to read files asynchronously (defaults to fs.readFile) */
-    readFile?: never;
+  } | {
     /** function to asynchronously read and parse a package.json file */
     readPackage?: (readFile: (file: string, cb: readFileCallback) => void, pkgfile: string, cb: readPackageCallback) => void;
-  }
-
-  export type AsyncOpts = AsyncOptsWithReadFile | AsyncOptsWithReadPackage;
+  });
 
   interface BaseSyncOpts extends Opts {
     /** function to synchronously test whether a file exists */
@@ -147,21 +139,13 @@ declare namespace resolve {
     realpathSync?: (file: string) => string;
   }
 
-  interface SyncOptsWithReadFile extends BaseSyncOpts {
+  export type SyncOpts = BaseSyncOpts & ({
     /** how to read files synchronously (defaults to fs.readFileSync) */
     readFileSync?: (file: string) => string | Buffer;
-    /** function to synchronously read and parse a package.json file */
-    readPackageSync?: never;
-  }
-
-  interface SyncOptsWithReadPackage extends BaseSyncOpts {
-    /** how to read files synchronously (defaults to fs.readFileSync) */
-    readFileSync?: never;
+  } | {
     /** function to synchronously read and parse a package.json file */
     readPackageSync?: (readFileSync: (file: string) => string | Buffer, pkgfile: string) => unknown;
-  }
-
-  export type SyncOpts = SyncOptsWithReadFile | SyncOptsWithReadPackage;
+  });
 
   export var sync: typeof resolveSync;
   export var isCore: typeof resolveIsCore;
```

with that no error is reported when both are specified. I guess I can keep the `never`

Reply:
yeah i'd still expect the never - but it seems like a composition-based approach (types) versus an inheritance-based approach (interfaces) will work better as more mutually exclusive options are added in the future.